squirrel.frontier/src/main/java/org/dice_research/squirrel/frontier/impl/FrontierImpl.java

```diff
@@ -267,15 +278,25 @@ public void run() {
 
     @Override
     public void addNewUris(List<CrawleableUri> uris) {
+        domainUriGenerator = new DomainBasedUriGenerator();
+        wellKnownUriGenrator = new WellKnownPathUriGenerator();
+        CrawleableUri domainUri, wellKnownUri;
         for (CrawleableUri uri : uris) {
+            // Generate the domain variant of the URI
+            domainUri = domainUriGenerator.getUriVariant(uri);
+            // Generate the well-known variant of a URI
+            wellKnownUri = wellKnownUriGenrator.getUriVariant(uri);
+            // Normalize the URI
+            uri = normalizer.normalize(uri);
             addNewUri(uri);
+            addNewUri(domainUri);
+            addNewUri(wellKnownUri);
+
         }
     }
```

Comment:
This solution is not bad. However, it has the disadvantage that the normalization and generation steps are only applied to URIs that are given as List while the simpler `addNewUri` method does not offer this functionality. However, the Frontier interface defines both as public and it does not define this difference.

The solution we discussed last time was to:
1. put the normlization and the generation into the `addNewUri` method (similar to the previous solution you had)
2. Move the if-else statement (marked here: https://github.com/CatherineChiramel/Squirrel/commit/cea6d7a11e5cb5f8615fe95f5ad34db4bbbc10c0#diff-412a8db07f54fa4995925670d4f53f95R303-R327) into an own method that is called with the normalized and the newly created URIs. The method should look like
```java
protected void addNormalizedUri(CrawleableUri uri) {
    ...
}
```